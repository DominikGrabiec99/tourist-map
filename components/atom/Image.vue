<template>
  <img class="image" :src="src" :alt="altText" />
</template>

<script lang="ts" setup>
defineProps({
  src: {
    type: String,
    defaul: '',
  },
  altText: {
    type: String,
    default: 'ZdjÄ™cie',
  },
});
</script>