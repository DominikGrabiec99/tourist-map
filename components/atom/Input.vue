<template>
  <div class="md:max-w-xs">
    <label
      v-if="label"
      :for="id"
      class="after:content-['*'] after:ml-0.5 after:text-red-500 block text-sm font-medium text-slate-700"
      :class="{
        'after:content-[]': !isRequire,
      }"
      >{{ label }}</label
    >
    <input
      :name="name"
      v-model="modelValue"
      :placeholder="placeholder"
      :id="id"
      :type="type"
      :disabled="disabled"
      :pattern="pattern"
      autocomplete="off"
      :class="{ 'border-red-500': error }"
      class="placeholder:italic placeholder:text-slate-400 mt-1 px-3 py-2 bg-white border shadow-sm border-slate-300 placeholder-slate-400 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1"
      @focus="emit('toggle-focus-input', true)"
      @blur="emit('toggle-focus-input', false)"
    />
    <label class="block text-sm font-medium text-red-500">
      {{ error }}
    </label>
  </div>
</template>

<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    placeholder?: string;
    id?: string;
    label?: string;
    isRequire?: boolean;
    name?: string;
    disabled?: boolean;
    error?: string;
    type?: string;
    pattern?: string;
  }>(),
  {
    placeholder: '',
    id: '',
    label: '',
    isRequire: false,
    name: '',
    disabled: false,
    error: '',
    type: 'text',
    pattern: '',
  }
);

const modelValue = defineModel<string>();

const emit = defineEmits<{
  (e: 'update:modelValue', value: string): void;
  (e: 'toggle-focus-input', value: boolean): void;
}>();
</script>
