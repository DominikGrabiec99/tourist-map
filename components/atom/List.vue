<template>
  <div class="list">
    <slot name="top" />
    <component
      :is="enumerateList ? 'ol' : 'ul'"
      class="list list-outside list-decimal"
      :class="{
        'list-inside list-disc': !enumerateList,
        'list-none': none,
        'list-square': square,
      }"
    >
      <li
        v-for="(item, index) in list?.length ? list : Object.keys($slots)"
        :key="`${listName}-item-${index}`"
        class="list-item"
        :class="liClass"
      >
        <slot :item="item" :index="index">
          <span v-html="item" />
        </slot>
      </li>
    </component>
    <slot name="bottom" />
  </div>
</template>

<script setup lang="ts">
const props = defineProps({
  listName: {
    type: String,
    require: true,
  },
  list: {
    type: Array,
    default: () => [],
  },
  enumerateList: {
    type: Boolean,
    default: false,
  },
  none: {
    type: Boolean,
    default: false,
  },
  square: {
    type: Boolean,
    default: false,
  },
  liClass: {
    type: String,
    require: '',
  },
});
</script>
