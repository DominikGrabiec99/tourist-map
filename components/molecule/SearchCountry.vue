<template>
  <div class="flex md:justify-end relative">
    <AtomInput
      v-model="modelValue"
      placeholder="ZnajdÅº kraj.."
      id="find-country"
      class="w-full"
      @toggle-focus-input="emit('toggle-focus-input', $event)"
    />
    <Transition>
      <div
        v-if="showResult"
        class="absolute bottom-0 right-0 w-full md:max-w-xs z-10 translate-y-full rounded-md border border-slate-300 dark:border-white transition-color"
      >
        <AtomList
          listName="search-country-list"
          :list="foundedCounties"
          none
          li-class="bg-slate-100 odd:bg-slate-200 dark:text-white dark:bg-neutral-600 dark:odd:bg-neutral-700 transition-color"
        >
          <template v-slot="{ item }">
            <span
              class="py-2 px-2 w-full block cursor-pointer"
              @click="emit('set-select-country', item.code as string)"
            >
              <AtomImage
                :src="`https://flagcdn.com/16x12/${item.code.toLowerCase()}.png`"
                :altText="item.name_pl"
                class="pr-1 pb-1 inline"
              />
              {{ item.name_pl }}
            </span>
          </template>
        </AtomList>
      </div>
    </Transition>
  </div>
</template>

<script setup lang="ts">
import IPolishCountryName from '@/ts/interfaces/PolishCountryName';

const props = withDefaults(
  defineProps<{
    modelValue?: string;
    foundedCounties?: IPolishCountryName[];
    showResult?: boolean;
  }>(),
  {
    foundedCounties: () => [],
    showResult: true,
  }
);

const modelValue = defineModel<string>();

const emit = defineEmits<{
  (e: 'toggle-focus-input', value: boolean): void;
  (e: 'set-select-country', value: string): void;
}>();
</script>

<style scoped lang="scss">
.v-enter-active,
.v-leave-active {
  transition: opacity 0.5s ease;
}

.v-enter-from,
.v-leave-to {
  opacity: 0;
}
</style>
