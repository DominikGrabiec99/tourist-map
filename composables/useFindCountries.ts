import ALL_COUNTRIES_WITH_POLISH_NAME from '@/constants/polish_name_countries'

const useFindCountries = () => {
  const getFoundedCounties = (searchCountryValue: string, searchAfterTextLength: number) => {
    if(searchCountryValue.length < searchAfterTextLength) return [];
    return ALL_COUNTRIES_WITH_POLISH_NAME.filter((country) => country.name_pl.toLowerCase().includes(searchCountryValue.toLowerCase()))
  }
  
  return {
    getFoundedCounties
  }
}

export default useFindCountries;