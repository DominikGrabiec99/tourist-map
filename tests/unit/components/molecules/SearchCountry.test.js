import { describe, it } from 'vitest';
import { mount } from '@vue/test-utils';

import SearchCountry from '@/components/molecule/SearchCountry.vue';
import Input from '@/components/atom/Input.vue';
import List from '@/components/atom/List.vue';
import Image from '@/components/atom/Image.vue';

import useFindCountries from '@/composables/useFindCountries.ts';

describe('SearchCountry', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(SearchCountry, {
      propsData: {
        modelValue: '',
        foundedCounties: [],
      },
      components: {
        AtomInput: Input,
        AtomList: List,
        AtomImage: Image,
      },
    });
  });

  it('render component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('toggle isFocusSearchCountryInput', async () => {
    expect(wrapper.vm.isFocusSearchCountryInput).toBeFalsy();
    await wrapper.vm.toggleFocusCountryInput(
      !wrapper.vm.isFocusSearchCountryInput
    );
    expect(wrapper.vm.isFocusSearchCountryInput).toBeTruthy();
  });

  it('should open countries list', async () => {
    const { getFoundedCounties } = useFindCountries();
    await wrapper.setProps({
      modelValue: 'Pol',
      foundedCounties: getFoundedCounties('Pol', 3),
    });
    await wrapper.vm.toggleFocusCountryInput(
      !wrapper.vm.isFocusSearchCountryInput
    );
    expect(wrapper.vm.showResult).toBeTruthy();
    expect(wrapper.text()).toContain(getFoundedCounties('Pol', 3)[0].name_pl);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render image', async () => {
    const { getFoundedCounties } = useFindCountries();
    await wrapper.setProps({
      modelValue: 'Pol',
      foundedCounties: getFoundedCounties('Pol', 3),
    });
    await wrapper.vm.toggleFocusCountryInput(
      !wrapper.vm.isFocusSearchCountryInput
    );
    const imageCountry = wrapper.get(
      `[data-test="image-country-${getFoundedCounties('Pol', 3)[0].code}"]`
    );
    expect(imageCountry.element.getAttribute('src')).toBe(
      'https://flagcdn.com/16x12/pf.png'
    );
  });
});
