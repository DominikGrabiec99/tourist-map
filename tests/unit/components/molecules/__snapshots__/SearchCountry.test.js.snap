// Vitest Snapshot v1

exports[`SearchCountry > render component 1`] = `
<div
  class="flex md:justify-end relative"
  data-v-5f819f67=""
>
  <div
    class="md:max-w-xs w-full"
    data-v-5f819f67=""
  >
    <!--v-if-->
    <input
      autocomplete="off"
      class="placeholder:italic placeholder:text-slate-400 mt-1 px-3 py-2 bg-white border shadow-sm border-slate-300 placeholder-slate-400 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1 dark:bg-neutral-800 dark:text-white outline-none focus:border-inherit focus:shadow-none transition-color"
      data-test="input"
      id="find-country"
      name=""
      pattern=""
      placeholder="Znajdź kraj.."
      type="text"
    />
    <!--v-if-->
  </div>
  <transition-stub
    appear="false"
    css="true"
    data-v-5f819f67=""
    persisted="false"
  >
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`SearchCountry > should open countries list 1`] = `
<div
  class="flex md:justify-end relative"
  data-v-5f819f67=""
>
  <div
    class="md:max-w-xs w-full"
    data-v-5f819f67=""
  >
    <!--v-if-->
    <input
      autocomplete="off"
      class="placeholder:italic placeholder:text-slate-400 mt-1 px-3 py-2 bg-white border shadow-sm border-slate-300 placeholder-slate-400 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1 dark:bg-neutral-800 dark:text-white outline-none focus:border-inherit focus:shadow-none transition-color"
      data-test="input"
      id="find-country"
      name=""
      pattern=""
      placeholder="Znajdź kraj.."
      type="text"
    />
    <!--v-if-->
  </div>
  <transition-stub
    appear="false"
    css="true"
    data-v-5f819f67=""
    persisted="false"
  >
    <div
      class="absolute bottom-0 right-0 w-full md:max-w-xs z-10 translate-y-full rounded-md border border-slate-300 dark:border-white transition-color"
      data-v-5f819f67=""
    >
      <div
        class="list"
        data-v-5f819f67=""
      >
        
        
        <ul
          class="list list-outside list-decimal list-inside list-disc list-none"
          data-test="list"
        >
          
          <li
            class="list-item bg-slate-100 odd:bg-slate-200 dark:text-white dark:bg-neutral-600 dark:odd:bg-neutral-700 transition-color"
          >
            
            <span
              class="py-2 px-2 w-full block cursor-pointer"
              data-test="element-list-span"
              data-v-5f819f67=""
            >
              <img
                alt="Polinezja Francuska"
                class="image pr-1 pb-1 inline"
                data-test="image-country-PF"
                data-v-5f819f67=""
                src="https://flagcdn.com/16x12/pf.png"
              />
               Polinezja Francuska
            </span>
            
          </li>
          <li
            class="list-item bg-slate-100 odd:bg-slate-200 dark:text-white dark:bg-neutral-600 dark:odd:bg-neutral-700 transition-color"
          >
            
            <span
              class="py-2 px-2 w-full block cursor-pointer"
              data-test="element-list-span"
              data-v-5f819f67=""
            >
              <img
                alt="Polska"
                class="image pr-1 pb-1 inline"
                data-test="image-country-PL"
                data-v-5f819f67=""
                src="https://flagcdn.com/16x12/pl.png"
              />
               Polska
            </span>
            
          </li>
          
        </ul>
        
        
      </div>
    </div>
  </transition-stub>
</div>
`;
